test/integration/http_timeout_integration_test.cc

```diff
@@ -499,4 +501,59 @@ void HttpTimeoutIntegrationTest::testRouterRequestAndResponseWithHedgedPerTryTim
   EXPECT_EQ("200", response->headers().getStatusValue());
 }
 
+// Starts a request with a header timeout specified, sleeps for longer than the
+// timeout, and ensures that a timeout is received.
+TEST_P(HttpTimeoutIntegrationTest, RequestHeaderTimeout) {
+  config_helper_.addConfigModifier(
+      [&](envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager&
+              hcm) {
+        auto* request_headers_timeout = hcm.mutable_request_headers_timeout();
+        request_headers_timeout->set_seconds(1);
+        request_headers_timeout->set_nanos(0);
+      });
+  setDownstreamProtocol(Http::CodecClient::Type::HTTP1);
+  initialize();
+
+  auto raw_connection = makeClientConnection(lookupPort("http"));
+  raw_connection->connect();
+
+  Buffer::OwnedImpl send_buffer("GET / HTTP/1.1\r\n"
+                                // Omit trailing \r\n that would indicate the end of headers.
+                                "Host: localhost\r\n");
+  // Track locally queued bytes, to make sure the outbound client queue doesn't back up.
+  uint64_t bytes_to_send = send_buffer.length();
+  raw_connection->addBytesSentCallback([&](uint64_t bytes) { bytes_to_send -= bytes; });
+  raw_connection->write(send_buffer, false);
```

Comment:
I think you can replace most of this with a call to
sendRawHttpAndWaitForResponse()